app/src/main/java/io/gnosis/safe/ui/safe/SafeSelectionDialog.kt

```diff
@@ -0,0 +1,277 @@
+    )
+
+    fun loadSafes() {
+        safeLaunch {
+            val safes = safeRepository.getSafes()
+            activeSafe = safeRepository.getActiveSafe()
+
+            items.clear()
+            items.add(AddSafeHeader())
+            items.addAll(safes)
+
+            updateState { SafeSelectionState.SafeListState(items, activeSafe, null) }
+        }
+    }
+
+    fun selectSafe(safe: Safe) {
+        safeLaunch {
+            safeRepository.setActiveSafe(safe)
+            updateState { SafeSelectionState.SafeListState(items, safe, null) }
+        }
+    }
+
+    fun addSafe() {
+        safeLaunch {
+            updateState {
+                SafeSelectionState.AddSafeState(
+                    ViewAction.NavigateTo(
+                        SafeOverviewFragmentDirections.actionSafeOverviewFragmentToAddSafeNav()
+                    )
+                )
+            }
+        }
+    }
+
+    override fun onSafeClicked(safe: Safe) {
+        selectSafe(safe)
+    }
+
+    override fun onAddSafeClicked() {
+        addSafe()
+    }
+}
+
+class SafeSelectionAdapter(
+    private val clickListener: WeakReference<OnSafeSelectionItemClickedListener>
+) : RecyclerView.Adapter<BaseSafeSelectionViewHolder>() {
+
+    private val items = mutableListOf<Any>()
+
+    var activeSafe: Safe? = null
+        set(value) {
+            field = value
+            notifyAllChanged()
+        }
+
+    fun setItems(items: List<Any>, activeSafe: Safe?) {
+        this.activeSafe = activeSafe
+        this.items.clear()
+        this.items.addAll(items)
+        notifyDataSetChanged()
+    }
+
+    override fun onBindViewHolder(holder: BaseSafeSelectionViewHolder, position: Int) {
+        when (holder) {
+            is AddSafeHeaderViewHolder -> holder.bind()
+            is SafeItemViewHolder -> {
+                val safe = items[position] as Safe
+
+                holder.bind(safe, safe == activeSafe)
+
+            }
+        }
+    }
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseSafeSelectionViewHolder {
+        return when (SafeSelectionViewTypes.values()[viewType]) {
+            SafeSelectionViewTypes.HEADER_ADD_SAFE -> AddSafeHeaderViewHolder(
+                ItemAddSafeBinding.inflate(
+                    LayoutInflater.from(parent.context),
+                    parent,
+                    false
+                ), clickListener
+            )
+            SafeSelectionViewTypes.SAFE -> SafeItemViewHolder(
+                ItemSafeBinding.inflate(LayoutInflater.from(parent.context), parent, false),
+                clickListener
+            )
+        }
+    }
+
+    override fun getItemViewType(position: Int): Int {
+        val item = items[position]
+        return when (item) {
+            is AddSafeHeader -> SafeSelectionViewTypes.HEADER_ADD_SAFE.ordinal
+            is Safe -> SafeSelectionViewTypes.SAFE.ordinal
+            else -> throw UnsupportedViewType(item.toString())
```

Comment:
Even though we are no using the super generic solution, defining a ViewHolder factory would make your adapter easier to digest.

Reply:
Same goes for `OnCreateViewHolder`

Reply:
I prefer not to do it in this pr